### ISSUE TITLE
[question] Why is the load balancing disabled for multi-mask models

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/torchtitan#3368]
I noticed that the PTRR load balancer is currently disabled for models using multiple attention masks. In `context_parallel.py`, there is an assertion enforcing that `mask` must not be a dict:

https://github.com/pytorch/torchtitan/blob/a2a0d99e333b27fe314359ff9c9d918a3c5363f6/torchtitan/distributed/context_parallel.py?utm_source=chatgpt.com#L208-L228

However, support for dict-based masks appears to already exist later in the same file:
https://github.com/pytorch/torchtitan/blob/a2a0d99e333b27fe314359ff9c9d918a3c5363f6/torchtitan/distributed/context_parallel.py?utm_source=chatgpt.com#L243-L263

I ran a few experiments locally, and PTRR load balancing seems to work correctly with multiple attention masks as well (even with interleaved SWA).

Is the restriction intentional, or could the assertion be relaxed/removed?

**Comment by fegin** [comment]
@francesco-bertolotti _PTRRLoadBalancer class only accepts on mask. How did you verify multiple masks? The issue of _PTRRLoadBalancer not supporting multiple masks is the algorithm itself. The algorithm doesn't take into account different masks may have different access patterns.

**Comment by francesco-bertolotti** [comment]
I used a custom model with interleaved sliding-window and full attention (similar to GPT-OSS). I passed a full block-causal attention mask to the _PTRRLoadBalancer, so the load balancing is driven entirely by the full-attention layers.

I ran a fairly quick, simple experiment, so I may be missing something, but the results seemed to match the loss observed when disabling the load balancer entirely.

Does this reasoning make sense?

**Comment by francesco-bertolotti** [comment]
@tianyu-l, I don’t think this is a bug---I was just trying to understand why the dict-masks are sharded even though they’re excluded from the load balancer.

I’m not very familiar with context parallelism and the load balancer, but it seems like the PTRR balancer works fine with the sliding window mask (though I could be mistaken on this).

**Comment by francesco-bertolotti** [comment]
I add that I modified `context_parallel.py` to avoid the assertions on the dict

**Comment by fegin** [comment]
@francesco-bertolotti Does the loss curve match? Since you have two attention masks, both need to be modified independently so that the each attention will attend the correct context. If you only passed one mask to PTRR, only that will be modified.

**Comment by francesco-bertolotti** [comment]
> @francesco-bertolotti Does the loss curve match?

I need to double-check what I did, but assuming I didn’t make any mistakes, the loss curve matched.

> Since you have two attention masks, both need to be modified

Yes, the load balancer is created from the full attention mask:

```python
load_balancer = _PTRRLoadBalancer(attention_masks["basic_mask"], cp_world_size)
```

And both masks  (the full attention mask and the SWA mask) are collected here:
https://github.com/pytorch/torchtitan/blob/a2a0d99e333b27fe314359ff9c9d918a3c5363f6/torchtitan/distributed/context_parallel.py#L244-L249

Both masks are then sharded here:
https://github.com/pytorch/torchtitan/blob/a2a0d99e333b27fe314359ff9c9d918a3c5363f6/torchtitan/distributed/context_parallel.py#L250-L255

**Comment by fegin** [comment]
@francesco-bertolotti I understand now. You are right that this will work as they share the same load balancer. However, my original statement holds, PTRR is designed to handle one single BlockMask, and with multiple masks, the load balancing become suboptimal.

**Comment by francesco-bertolotti** [comment]
> I understand now. You are right that this will work as they share the same load balancer.

That is great to know since we saw some speedup

> However, my original statement holds, PTRR is designed to handle one single BlockMask, and with multiple masks, the load balancing become suboptimal.

Yep, I imagined it could be suboptimal but I figured it couldn't be too bad since the other mask (being a SWA mask) is essentially a sub-mask of the full one.

If supporting something like this is interesting to you I can submit a PR.

**Comment by fegin** [comment]
I'm curious about your thought on how to decide the "main" mask, this will lead to a different load balancing result.

**Comment by francesco-bertolotti** [comment]
I don’t think there’s a single correct answer here. The optimal mask for load balancing will likely depend on the specific setup. For example, the number of layers using the mask, whether sequences are long or short, and similar factors.

Because of that, the cleanest approach is probably to let the user decide which mask should be used for load balancing.

For example, instead of passing `attention_masks` directly to the load balancer, we could do something like:

```python
if isinstance(attention_masks, dict):
    attention_masks = {"basic_mask": attention_masks}

load_balancer = _PTRRLoadBalancer(
    attention_masks.get(
        "load_balancing_mask",
        attention_masks["basic_mask"],
    ),
    cp_world_size,
)
```
